### PR TITLE
Adds support to autodiscover multiple memcached clusters with different topologies

### PR DESCRIPTION
**Pull request by miriampena**
We add support the setup of multiple pyhisical clusters with autodiscovery assigned to the same logical cluster. It could be the case in which some of these pyhisical clusters perform better than others, in which you can use a third argument called the ClusterSpeedFactor which has to be a small integer. The default ClusterSpeedFactor is 1.

If an alternate cluster is 2 times faster than the fist one -it can have twice as many cpus or memory-, you can set it up
with a ClusterSpeedFactor of 2. This will create 2 times more workers for that "faster cluster", which in practice will send twice as many connections & requests to that cluster than to the other weaker clusters.

```
    [{cluster_c,
        [{servers,
           {elasticache,
            [{"ConfigEndpointHostname.com", 11211},
             {"ConfigEndpointHostnameAltThatPerforms2TimesBetter.com", 11211, 2},
             {"ConfigEndpointHostnameAltThatPerforms3TimesBetter.com", 11211, 3}]
            }},
         {sharding_algorithm, {mero, shard_crc32}},
         {workers_per_shard, 1},
         {pool_worker_module, mero_wrk_tcp_binary}]}]
```

Contains https://github.com/AdRoll/mero/pull/29

### CODE REVIEW

**Comment by miriampena** [comment]
Do not merge yet. Going to test deploy it first.

File: src/mero_elasticache.erl [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        || Name <- Names],
+    ok = gen_tcp:close(Socket),
+    Lines.
+
+%% Parse host and version lines to return version and list of {host, port} cluster nodes
+-spec parse_cluster_config(binary()) -> [{string(), integer()}].
+parse_cluster_config(HostLine) ->
+    HostSpecs = re:split(butlast(HostLine), <<" ">>),
+    [begin
+         [Host, _IP, Port] = re:split(HIP, "\\|"),
```
[the remark this answers is not in the document]
Comment (zerth) [reply src/mero_elasticache.erl]: We only do discovery at init time, and the IP address of a replacement node could be different (in EC2-classic).

File: src/mero_elasticache.erl [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+         {binary_to_list(Host), binary_to_integer(Port)}
+     end
+        || HIP <- HostSpecs].
+
+butlast(<<>>) -> <<>>;
+butlast(Bin) -> binary:part(Bin, {0, size(Bin) - 1}).
+
+gen_tcp_send(Socket, Command) ->
+    gen_tcp:send(Socket, Command).
```
[the remark this answers is not in the document]
Comment (miriampena) [reply src/mero_elasticache.erl]: I did for debugging purposes. Ill remove it. it is not necessary anymore

File: src/mero_conf.erl [hunk 3]
```diff
@@ -217,3 +223,33 @@ get_env(Key) ->
         undefined ->
             exit({undefined_configuration, Key})
     end.
+
+
+process_value({servers, {elasticache, ConfigEndpoint, ConfigPort}}) ->
+    process_value({servers, {elasticache, [{ConfigEndpoint, ConfigPort, 1}]}});
+process_value({servers, {elasticache, ConfigList}}) when is_list(ConfigList) ->
+    HostsPorts = lists:foldr(
```
Comment (elbrujohalcon) [inline src/mero_conf.erl]: What do you think about reducing this huge `foldr` to something like the following (not tested on a console, just a concept) code?
```erlang
    HostsPorts = [Port || HostConfig <- ConfigList, Port <- get_ports(HostConfig)].
```
…and then somewhere else…
```erlang
get_ports({HostIn, PortIn}) ->
    get_ports({HostIn, PortIn, ClusterSpeedFactorIn});
get_ports({HostIn, PortIn, ClusterSpeedFactorIn}) ->
    lists:duplicate(ClusterSpeedFactor, get_elasticache_cluster_config(Host, Port)).
```

(you probably have to choose better names for this)

File: src/mero_elasticache.erl [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    parse_cluster_entries(Entries1, []).
+
+parse_cluster_entries([], Accum) ->
+    {ok, lists:reverse(Accum)};
+parse_cluster_entries([H|T], Accum) ->
+    case string:tokens(H, "|") of
+        [Host, IP, Port] ->
+            case inet:parse_ipv4_address(IP) of
+                {ok, IPAddr} ->
+                    case catch erlang:list_to_integer(Port) of
```
Comment (elbrujohalcon) [inline src/mero_elasticache.erl]: You should consider using `string:to_integer/1` instead of `catch erlang:list_to_integer/1` here.

**Pull request merged**